File: 207/solution.go
func canFinish(numCourses int, prerequisites [][]int) bool {
	// Initialized Value
	adj := make([][]int, numCourses)
	inDegreeNum := make([]int, numCourses)
	q := make([]int, 0)

	for i := 0; i < numCourses; i++ {
		inDegreeNum[i] = 0
	}

	// Build Adjency Matrix
	for i := 0; i < len(prerequisites); i++ {
		pair := prerequisites[i]
		adj[pair[1]] = append(adj[pair[1]], pair[0])
		inDegreeNum[pair[0]]++
	}

	// Find Starting Points
	for i := 0; i < numCourses; i++ {
		if inDegreeNum[i] == 0 {
			q = append(q, i)
		}
	}

	for len(q) > 0 {
		// Get first Element
		candidate := q[0]
		q = q[1:]
		numCourses--

		// Traverse
		for _, val := range adj[candidate] {
			inDegreeNum[val]--
			if inDegreeNum[val] == 0 {
				q = append(q, val)
			}
		}
	}

	return numCourses == 0
}

File: 207/solution.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj(numCourses, vector<int>());
        vector<int> inDegreeNum(numCourses, 0);
        queue<int> q;
        
        // Build adjency information
        for (auto &prerequisite: prerequisites) {
            adj[prerequisite[1]].push_back(prerequisite[0]);
            inDegreeNum[prerequisite[0]]++;
        }
    
        // Find starting point
        for (int i=0; i<numCourses; i++) {
            if (inDegreeNum[i] == 0) q.push(i);
        }
        
        
        // Topology sort
        while(!q.empty()) {
            // Get the first element
            int candidate = q.front();
            q.pop();
            numCourses--;
            
            // Traverse first element
            for (auto linkedElement: adj[candidate]) {
                if (--inDegreeNum[linkedElement] <= 0) q.push(linkedElement);
            }
        }
        
        return numCourses == 0;
    }
};